src/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/XmlWriterTraceListener.cs

```diff
@@ -0,0 +1,375 @@
+﻿using System.Collections;
+using System.Globalization;
+using System.IO;
+using System.Runtime.Versioning;
+using System.Text;
+using System.Threading;
+using System.Xml;
+using System.Xml.XPath;
+
+namespace System.Diagnostics
+{
+    public class XmlWriterTraceListener : TextWriterTraceListener
+    {
+        private const string FixedHeader = "<E2ETraceEvent xmlns=\"http://schemas.microsoft.com/2004/06/E2ETraceEvent\"><System xmlns=\"http://schemas.microsoft.com/2004/06/windows/eventlog/system\">";
+
+        private readonly string _machineName = Environment.MachineName;
+        private StringBuilder _strBldr = null;
+        private XmlTextWriter _xmlBlobWriter = null;
+
+        public XmlWriterTraceListener(Stream stream)
+            : base(stream)
+        {
+        }
+
+        public XmlWriterTraceListener(Stream stream, string name)
+            : base(stream, name)
+        {
+        }
+
+        public XmlWriterTraceListener(TextWriter writer)
+            : base(writer)
+        {
+        }
+
+        public XmlWriterTraceListener(TextWriter writer, string name)
+            : base(writer, name)
+        {
+        }
+
+        [ResourceExposure(ResourceScope.Machine)]
+        [ResourceConsumption(ResourceScope.Machine)]
+        public XmlWriterTraceListener(string filename)
+            : base(filename)
+        {
+        }
+
+        [ResourceExposure(ResourceScope.Machine)]
+        [ResourceConsumption(ResourceScope.Machine)]
+        public XmlWriterTraceListener(string filename, string name)
+            : base(filename, name)
+        {
+        }
+
+        public override void Write(string message)
+        {
+            WriteLine(message);
+        }
+
+        public override void WriteLine(string message)
+        {
+            TraceEvent(null, SR.TraceAsTraceSource, TraceEventType.Information, 0, message);
+        }
+
+        public override void Fail(string message, string detailMessage)
+        {
+            int length = detailMessage != null ? message.Length + 1 + detailMessage.Length : message.Length;
+            TraceEvent(null, SR.TraceAsTraceSource, TraceEventType.Error, 0, string.Create(length, detailMessage,
+            (dst, v) =>
+            {
+                ReadOnlySpan<char> prefix = message;
```

Comment:
I assume allocation in this code doesn't really matter, but this change has kind of left it in a strange middle-ground, as some allocations are being avoided but not as many as could be, e.g. this is now resulting in an allocation for a delegate and for a closure because it's capturing message.  You can instead pass in both message and detailMessage in a value tuple, e.g. `(message, detailMessage)`, and use the values passed into the delegate rather than capturing one of them.

Reply:
Good point, I didn't came up with the tuple idea when was writing this code. Thanks for the advice :smile: